File: scripts/rank.py
"""Rank investors for a mandate, and say why.

Two families of signal, and the distinction matters more than the arithmetic:

  STATED    what the fund says about itself — strategy, cheque range, geography.
            Covers all 1,285 investors, so it is comparable across the universe.

  REVEALED  who it has actually backed — the portfolio companies the crawler
            found, their names and descriptions, and how recently.
            Covers ~45 investors today. That is 3.5% of the universe.

**Past engagement is deliberately not a signal.** The `prior_approach` value
exists in the match_signal enum and is not used here. A fund passing on a
mandate reflects its timing, fund cycle or a portfolio conflict far more often
than fit, so scoring on it would penalise the busiest and most relevant houses.
The ICS outcomes are for checking this ranking, never for building it.

The coverage gap forces one design decision. If revealed fit simply added to a
total, the ranking would put the houses we happen to have crawled on top — an
artefact of the crawl schedule, not a judgement about fit. So the two families
are scored and displayed **separately**: stated fit ranks the universe, revealed
fit is shown as corroboration where it exists, and an investor is never marked
down for having no portfolio data yet.
"""

from __future__ import annotations

import argparse
import json
import os
import re
import sys
from collections import Counter, defaultdict

sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

from crawler import db      # noqa: E402

VERSION = "1"

WEIGHTS = {
    "sector_fit": 0.35,
    "size_fit": 0.25,
    "deal_type_fit": 0.20,
    "geography_fit": 0.20,
}
REVEALED_WEIGHTS = {
    "thesis_similarity": 0.65,
    "recent_activity": 0.35,
}

# Which strategies a mandate type implies. A fund can hold several.
TYPE_TO_STRATEGY = {
    "growth_capital": {"growth", "venture"},
    "sell_side": {"buyout", "growth"},
    "buy_side": {"buyout"},
    "debt": {"credit", "mezzanine"},
    "lp_fundraise": {"fund_of_funds", "secondaries"},
    "continuation_fund": {"secondaries", "buyout"},
    "secondary": {"secondaries"},
    "strategic_review": {"buyout", "growth"},
}

# A mandate is written in a banker's vocabulary and the investor's sector focus
# comes from Crunchbase's. "Retail technology" and "Commerce and Shopping" mean
# the same thing and share no words, so both sides are mapped to concepts before
# comparison. Deliberately small and explicit — a large synonym table would be
# unauditable, and a wrong bridge is worse than a missing one.
CONCEPTS = {
    "commerce": {"retail", "retailer", "retailers", "consumer", "shopping",
                 "commerce", "ecommerce", "e-commerce", "checkout", "grocery",
                 "store", "stores", "merchandising", "brand", "brands", "fashion"},
    "software": {"software", "saas", "platform", "application", "applications",
                 "apps", "cloud"},
    "technology": {"technology", "information", "digital", "tech"},
    "internet": {"internet", "online", "marketplace", "services"},
    "data": {"data", "analytics", "artificial", "intelligence", "machine",
             "vision", "learning"},
    "fintech": {"fintech", "payments", "payment", "lending", "banking",
                "insurance", "insurtech", "wealth"},
    "health": {"health", "healthcare", "medical", "clinical", "pharma",
               "biotechnology", "biotech", "diagnostics", "care"},
    "media": {"media", "entertainment", "content", "publishing", "gaming"},
    "industrial": {"manufacturing", "industrial", "engineering", "science",
                   "materials", "chemicals"},
    "energy": {"energy", "renewable", "cleantech", "climate", "sustainability"},
    "property": {"property", "estate", "proptech", "construction", "housing"},
    "transport": {"transport", "transportation", "logistics", "mobility",
                  "automotive", "supply"},
    "mobile": {"mobile", "telecom", "telecoms", "wireless", "connectivity"},
    "security": {"security", "privacy", "cyber", "identity"},
    "hardware": {"hardware", "electronics", "devices", "semiconductor", "robotics"},
    "commercial": {"sales", "marketing", "recruitment", "staffing", "education"},
}


def concepts(words: set[str]) -> set[str]:
    return {c for c, vocab in CONCEPTS.items() if words & vocab}


STOPWORDS = {
    "the", "and", "for", "with", "that", "this", "from", "into", "over", "under",
    "company", "companies", "business", "businesses", "group", "limited", "ltd",
    "plc", "uk", "market", "markets", "leading", "provider", "providers",
    "services", "service", "solutions", "solution", "platform", "based",
    "products", "product", "customers", "clients", "million", "growth",
}


def keywords(text: str) -> set[str]:
    words = re.findall(r"[a-z][a-z0-9\-]{3,}", (text or "").lower())
    return {w for w in words if w not in STOPWORDS}


def mandate(conn, code: str) -> dict | None:
    return db.one(
        conn,
        """
        select m.id, m.code, m.name, m.mandate_type, m.country_code,
               m.revenue_gbp, m.ebitda_gbp, m.expected_ev_gbp,
               m.business_description, m.hard_filters,
               coalesce(m.required_country_codes,
                        array[m.country_code]::char(2)[]) as required_countries,
               s.name as sector
        from public.mandates m
        left join public.sectors s on s.id = m.sector_id
        where m.code = %s
        """,
        (code,),
    )


def candidates(conn) -> list[dict]:
    return db.all_rows(
        conn,
        """
        select i.company_id, c.legal_name, c.country_code,
               i.min_cheque_raw, i.max_cheque_raw, i.cheque_currency,
               array_remove(array_agg(distinct s.strategy::text), null) as strategies,
               array_remove(array_agg(distinct sec.name), null) as focus,
               max(gr.grade) as grade,
               bool_or(gr.never_approach) as never_approach
        from public.investors i
        join public.companies c on c.id = i.company_id
        left join public.v_investor_current_grade gr on gr.company_id = i.company_id
        left join public.investor_strategies s on s.company_id = i.company_id
        left join public.investor_sector_focus f on f.company_id = i.company_id
        left join public.sectors sec on sec.id = f.sector_id
        where c.merged_into_id is null
        group by i.company_id, c.legal_name, c.country_code,
                 i.min_cheque_raw, i.max_cheque_raw, i.cheque_currency
        """,
    )


def portfolios(conn) -> dict[str, dict]:
    """Per investor: the text of what it has backed, and how recently.

    Text comes from the claims rather than the promoted `companies` rows,
    because promotion writes only a name — the descriptions the crawler
    captured stay on the claim, and they are what makes a thesis match
    explainable rather than a bare name collision.
    """
    text_rows = db.all_rows(
        conn,
        """
        select subject_id,
               string_agg(coalesce(value_text, '') || ' ' ||
                          coalesce(value_json->>'description', ''), ' ') as blob,
               count(*) as holdings
        from public.claims
        where subject_table = 'investors' and attribute = 'portfolio_company'
        group by subject_id
        """,
    )
    recency = db.all_rows(
        conn,
        """
        select investor_company_id,
               max(extract(year from entry_date))::int as latest,
               count(*) filter (where entry_date > now() - interval '4 years') as recent
        from public.investments
        where entry_date is not null
        group by investor_company_id
        """,
    )
    out: dict[str, dict] = defaultdict(dict)
    for r in text_rows:
        out[str(r["subject_id"])] = {"blob": r["blob"] or "", "holdings": r["holdings"],
                                     "latest": None, "recent": 0}
    for r in recency:
        key = str(r["investor_company_id"])
        out.setdefault(key, {"blob": "", "holdings": 0})
        out[key]["latest"] = r["latest"]
        out[key]["recent"] = r["recent"]
    return out


def named_matches(blob: str, wanted: set[str], limit: int = 3) -> list[str]:
    """Which holdings matched, so the reason can name them rather than assert."""
    hits = []
    for line in re.split(r"(?<=[a-z])\s(?=[A-Z])|\n", blob):
        if len(hits) >= limit:
            break
        if keywords(line) & wanted:
            name = line.strip()[:48]
            if name and name not in hits:
                hits.append(name)
    return hits


def excluded_by(m: dict, inv: dict, wanted_concepts: set[str]) -> str | None:
    """Absolute disqualifications, checked before any scoring.

    A fund outside the mandate's geography is not a 40%-good answer, it is not
    an answer, and a strong sector score must not be able to drag it back in.
    Which criteria are absolute is the mandate's decision, not the engine's.

    Every exclusion is returned with its reason so the run can report what it
    threw away — a ranking that silently drops candidates is not auditable.
    """
    if inv.get("never_approach"):
        return "graded never-approach"

    hard = set(m.get("hard_filters") or [])
    if "geography" in hard:
        required = [c for c in (m.get("required_countries") or []) if c]
        if required:
            if not inv.get("country_code"):
                return "geography unknown"
            if inv["country_code"] not in required:
                return f"outside {'/'.join(required)}"

    if "sector" in hard and wanted_concepts:
        focus = [f for f in (inv.get("focus") or []) if f]
        if not focus:
            return "no stated sector focus"
        if not (concepts(keywords(" ".join(focus))) & wanted_concepts):
            return "sector mismatch"

    if "size" in hard:
        target = m.get("expected_ev_gbp") or m.get("revenue_gbp")
        lo, hi = inv.get("min_cheque_raw"), inv.get("max_cheque_raw")
        if target and lo is not None and hi is not None:
            if not (float(lo) <= float(target) <= float(hi)):
                return "cheque range excludes the mandate"

    return None


def score(m: dict, inv: dict, port: dict | None, wanted: set[str],
          wanted_concepts: set[str]) -> dict:
    stated, revealed, reasons = {}, {}, {}

    # -- stated sector focus ---------------------------------------------
    focus = [f for f in (inv.get("focus") or []) if f]
    if focus and wanted_concepts:
        theirs = concepts(keywords(" ".join(focus)))
        shared = wanted_concepts & theirs
        stated["sector_fit"] = len(shared) / len(wanted_concepts)
        if shared:
            matched = [f for f in focus if concepts(keywords(f)) & shared]
            reasons["sector_fit"] = f"states focus on {', '.join(matched[:3])}"

    # -- deal type -------------------------------------------------------
    expected = TYPE_TO_STRATEGY.get(m.get("mandate_type") or "sell_side", set())
    held = set(inv["strategies"] or [])
    if expected and held:
        overlap = held & expected
        stated["deal_type_fit"] = len(overlap) / len(expected) if overlap else 0.0
        if overlap:
            reasons["deal_type_fit"] = f"invests in {', '.join(sorted(overlap))}"

    # -- cheque size -----------------------------------------------------
    target = m.get("expected_ev_gbp") or m.get("revenue_gbp")
    lo, hi = inv.get("min_cheque_raw"), inv.get("max_cheque_raw")
    if target and (lo is not None or hi is not None):
        lo = float(lo or 0)
        hi = float(hi or lo * 10 or 0)
        target = float(target)
        if lo <= target <= hi:
            stated["size_fit"] = 1.0
            reasons["size_fit"] = f"writes {lo/1e6:.0f}–{hi/1e6:.0f}m, mandate at {target/1e6:.0f}m"
        else:
            # Distance in orders of magnitude, so being 2x out is not the same
            # as being 50x out.
            edge = lo if target < lo else hi
            ratio = max(target, edge) / max(min(target, edge), 1)
            stated["size_fit"] = max(0.0, 1 - (ratio - 1) / 4)

    # -- geography -------------------------------------------------------
    if m.get("country_code") and inv.get("country_code"):
        same = inv["country_code"] == m["country_code"]
        stated["geography_fit"] = 1.0 if same else 0.4
        if same:
            reasons["geography_fit"] = f"based in {inv['country_code']}"

    # -- revealed: what they have actually backed ------------------------
    if port and port.get("blob"):
        held_words = keywords(port["blob"])
        if wanted:
            shared = wanted & held_words
            revealed["thesis_similarity"] = min(1.0, len(shared) / max(len(wanted), 1) * 2)
            if shared:
                names = named_matches(port["blob"], wanted)
                reasons["thesis_similarity"] = (
                    f"portfolio overlaps on {', '.join(sorted(shared)[:5])}"
                    + (f" — e.g. {'; '.join(names)}" if names else "")
                )
    if port and port.get("recent") is not None:
        revealed["recent_activity"] = min(1.0, (port.get("recent") or 0) / 5)
        if port.get("latest"):
            reasons["recent_activity"] = (
                f"{port.get('recent', 0)} investments in the last 4 years, "
                f"most recent {port['latest']}")

    def blend(scores: dict, weights: dict) -> float:
        """Normalise over ALL weights, not just the ones we could compute.

        Dividing by the present weights treats a missing signal as "not
        applicable" and hands a perfect score to an investor whose entire case
        is being in the right country: Accelerator London scored 1.00 on
        geography alone and tied with Apax, which matched on strategy, cheque
        size and geography too. Missing data is unknown, not neutral, so an
        absent signal contributes nothing and completeness is rewarded.
        """
        if not scores:
            return 0.0
        return sum(scores.get(k, 0.0) * w for k, w in weights.items()) / sum(weights.values())

    return {
        "stated": round(blend(stated, WEIGHTS), 4),
        "revealed": round(blend(revealed, REVEALED_WEIGHTS), 4) if revealed else None,
        "signals": {**stated, **revealed},
        "reasons": reasons,
        "holdings": (port or {}).get("holdings", 0),
    }


def run(conn, code: str, top: int, apply: bool) -> int:
    m = mandate(conn, code)
    if not m:
        print(f"no mandate with code {code!r}")
        return 1

    wanted = keywords(" ".join(filter(None, [m.get("business_description"),
                                             m.get("sector"), m.get("name")])))
    print(f"  {m['name']}")
    print(f"  type={m['mandate_type']} country={m['country_code']} "
          f"ev={m['expected_ev_gbp']} sector={m['sector']}")
    wanted_concepts = concepts(wanted)
    print(f"  {len(wanted)} keywords -> concepts: {', '.join(sorted(wanted_concepts))}\n")

    port = portfolios(conn)
    rows = candidates(conn)
    scored, excluded = [], Counter()
    for inv in rows:
        reason = excluded_by(m, inv, wanted_concepts)
        if reason:
            excluded[reason] += 1
            continue
        s = score(m, inv, port.get(str(inv["company_id"])), wanted, wanted_concepts)
        if s["stated"] <= 0 and not s["revealed"]:
            continue
        scored.append({**inv, **s})

    if excluded:
        print(f"  hard filters ({', '.join(sorted(set(m.get('hard_filters') or [])))}) "
              f"excluded {sum(excluded.values())}:")
        for reason, n in excluded.most_common(6):
            print(f"      {reason:34s} {n}")
        print()

    # Stated fit ranks the universe; revealed fit breaks ties and corroborates.
    # Adding them would rank the crawl schedule, not the investors.
    scored.sort(key=lambda r: (-(r["stated"]), -(r["revealed"] or 0)))
    with_evidence = sum(1 for r in scored if r["revealed"] is not None)
    print(f"  {len(scored)} scoreable investors, {with_evidence} with portfolio evidence\n")

    for rank, r in enumerate(scored[:top], 1):
        ev = f"{r['revealed']:.2f}" if r["revealed"] is not None else "  — "
        grade = f"  grade {r['grade'].upper()}" if r.get("grade") else ""
        print(f"  {rank:2d}. {r['legal_name'][:34]:34s} stated {r['stated']:.2f}   "
              f"evidence {ev}  ({r['holdings']} holdings){grade}")
        for signal, why in list(r["reasons"].items())[:3]:
            print(f"        {signal:18s} {why[:96]}")

    if not apply:
        print("\n  dry run — nothing written. Add --apply to record the run.")
        return 0

    run_id = db.scalar(
        conn,
        """
        insert into public.match_runs
            (mandate_id, algorithm_version, parameters, candidate_count,
             returned_count, status, finished_at, is_current)
        values (%s, %s, %s, %s, %s, 'ok', now(), false)
        returning id
        """,
        (m["id"], VERSION,
         json.dumps({"stated": WEIGHTS, "revealed": REVEALED_WEIGHTS,
                     "engagement_used": False}),
         len(rows), min(top, len(scored))),
    )
    for rank, r in enumerate(scored[:top], 1):
        match_id = db.scalar(
            conn,
            """
            insert into public.investor_matches
                (run_id, investor_company_id, rank, total_score, score_breakdown,
                 headline_rationale)
            values (%s, %s, %s, %s, %s, %s)
            returning id
            """,
            (run_id, r["company_id"], rank, r["stated"],
             json.dumps({"stated": r["stated"], "revealed": r["revealed"],
                         "signals": r["signals"]}),
             "; ".join(list(r["reasons"].values())[:2])[:400]),
        )
        for signal, why in r["reasons"].items():
            weight = WEIGHTS.get(signal) or REVEALED_WEIGHTS.get(signal) or 0
            db.execute(
                conn,
                """
                insert into public.investor_match_reasons
                    (match_id, signal, raw_score, weight, narrative)
                values (%s, %s::public.match_signal, %s, %s, %s)
                on conflict (match_id, signal) do nothing
                """,
                (match_id, signal, r["signals"].get(signal, 0), weight, why[:500]),
            )
    conn.commit()
    print(f"\n  run {str(run_id)[:8]} recorded, {min(top, len(scored))} matches")
    return 0


def main() -> int:
    ap = argparse.ArgumentParser()
    ap.add_argument("--code", required=True, help="mandate code, e.g. ics-monument")
    ap.add_argument("--top", type=int, default=20)
    ap.add_argument("--apply", action="store_true")
    args = ap.parse_args()
    with db.connect() as conn:
        return run(conn, args.code, args.top, args.apply)


if __name__ == "__main__":
    sys.exit(main())

File: web/lib/rank.ts
// Scoring, ported from scripts/rank.py so there is one implementation rather
// than a Python engine drifting from a TypeScript UI. The design decisions it
// carries are the ones that took the longest to get right:
//
//   * Ardent's own descriptors beat anything inferred. Fund type, investment
//     geography, sector and cheque band come off the Investor Control Sheet,
//     which is maintained by hand for the purpose of deciding who to approach.
//     Crawled strategy tags and HQ country are the fallback, not the source.
//   * Stated fit and revealed fit are scored SEPARATELY and never summed.
//     Portfolio evidence exists for ~3.5% of investors; adding it to a total
//     would rank the crawl schedule rather than the investors.
//   * Missing signals score zero rather than being excluded from the average.
//     Normalising over present signals only gave a perfect score to anyone
//     merely based in the right country.
//   * Hard filters exclude on KNOWN failure only. Most investors have no
//     investment geography on file, and dropping them would remove a third of
//     the market from every UK mandate because of a gap in our own data.
//   * Past engagement is not a signal. A fund passes for timing or fund cycle
//     far more often than for fit, so scoring on declines would penalise the
//     busiest and most relevant houses.

export type Mandate = {
  // Recorded for the eventual report, no longer scored. It used to drive deal
  // type through a mandate-to-strategy map, which put "sell side" in a dropdown
  // above a list of investors — and there is no such thing as a sell-side
  // investor. Fund type is the investor's own attribute and is now chosen
  // directly.
  mandate_type: string | null;
  fund_types: string[];
  country_code: string | null;
  expected_ev_gbp: number | null;
  revenue_gbp: number | null;
  sector: string | null;
  business_description: string | null;
  hard_filters: string[];
  required_countries: string[];
};

export type Investor = {
  company_id: string;
  legal_name: string;
  country_code: string | null;
  fund_type: string | null;
  invest_geographies: string[];
  ardent_sector: string | null;
  check_band: string | null;
  cheque_min: number | null;
  cheque_max: number | null;
  cheque_source: string | null;
  strategies: string[];
  focus: string[];
  grade: string | null;
  never_approach: boolean;
  holdings: number;
  blob: string;
  recent: number;
  latest: number | null;
};

export const WEIGHTS: Record<string, number> = {
  sector_fit: 0.35,
  size_fit: 0.25,
  deal_type_fit: 0.2,
  geography_fit: 0.2,
};
const REVEALED: Record<string, number> = {
  thesis_similarity: 0.65,
  recent_activity: 0.35,
};

// Ardent's own fund categories, in the order the Investor Control Sheet lists
// them. This IS the vocabulary — nothing infers it, and the analyst edits it.
export const ARDENT_FUND_TYPES: [string, string][] = [
  ["hnw", "High Net Worth"],
  ["family_office", "Family Office"],
  ["vct", "EIS / VCT"],
  ["vc", "VC"],
  ["mbo", "MBO"],
  ["lbo", "LBO"],
  ["lp", "LP"],
  ["multi_strategy_am", "Multi-Strategy AM"],
  ["corp_strategy_office", "Corp Strategy Office"],
];

export function fundTypeLabel(key: string | null): string {
  if (!key) return "";
  return ARDENT_FUND_TYPES.find(([k]) => k === key)?.[1] || key.replace(/_/g, " ");
}

// Deal type is scored only when the analyst has named the fund types they want.
// Selecting none is not a missing signal about the investor — it is a question
// nobody asked — so the weight is removed rather than scored zero, which is the
// opposite of how a genuinely absent attribute is treated everywhere else here.
function weightsFor(m: Mandate): Record<string, number> {
  if (m.fund_types && m.fund_types.length) return WEIGHTS;
  const { deal_type_fit, ...rest } = WEIGHTS;
  void deal_type_fit;
  return rest;
}

// The sheet records where a fund DEPLOYS as one of four regions. A mandate
// arrives with an ISO country code, so it has to be lifted to the same grain.
// Not exhaustive by design: an unlisted country yields no region, which means
// no geography signal rather than a wrong one.
const COUNTRY_TO_REGION: Record<string, string> = {
  GB: "UK",
  IE: "EU", FR: "EU", DE: "EU", NL: "EU", BE: "EU", LU: "EU", ES: "EU", PT: "EU",
  IT: "EU", AT: "EU", CH: "EU", DK: "EU", SE: "EU", NO: "EU", FI: "EU", PL: "EU",
  CZ: "EU", GR: "EU",
  US: "NA", CA: "NA", MX: "NA",
  CN: "ASIA", HK: "ASIA", JP: "ASIA", SG: "ASIA", IN: "ASIA", KR: "ASIA",
  TW: "ASIA", MY: "ASIA", ID: "ASIA", AE: "ASIA", IL: "ASIA",
};

export function regionFor(country: string | null): string | null {
  return country ? COUNTRY_TO_REGION[country.toUpperCase()] || null : null;
}

// A mandate is written in a banker's vocabulary; investor sector focus comes
// from Crunchbase's. "Retail technology" and "Commerce and Shopping" mean the
// same thing and share no words. Kept small and explicit on purpose — an
// unauditable synonym table is worse than a missing bridge.
const CONCEPTS: Record<string, string[]> = {
  commerce: ["retail", "retailer", "retailers", "consumer", "shopping", "commerce",
    "ecommerce", "checkout", "grocery", "store", "stores", "brand", "brands", "fashion"],
  software: ["software", "saas", "platform", "application", "applications", "apps", "cloud"],
  technology: ["technology", "information", "digital", "tech"],
  internet: ["internet", "online", "marketplace"],
  data: ["data", "analytics", "artificial", "intelligence", "machine", "vision"],
  fintech: ["fintech", "payments", "payment", "lending", "banking", "insurance", "wealth"],
  health: ["health", "healthcare", "medical", "clinical", "pharma", "biotechnology", "care"],
  media: ["media", "entertainment", "content", "publishing", "gaming"],
  industrial: ["manufacturing", "industrial", "engineering", "science", "materials"],
  energy: ["energy", "renewable", "cleantech", "climate"],
  property: ["property", "estate", "proptech", "construction"],
  transport: ["transport", "transportation", "logistics", "mobility", "automotive"],
  mobile: ["mobile", "telecom", "telecoms", "wireless"],
  security: ["security", "privacy", "cyber"],
  hardware: ["hardware", "electronics", "devices", "semiconductor", "robotics"],
  commercial: ["sales", "marketing", "recruitment", "staffing", "education"],
};

const STOP = new Set(["the", "and", "for", "with", "that", "this", "from", "company",
  "companies", "business", "group", "limited", "market", "leading", "provider",
  "services", "service", "solutions", "platform", "products", "customers", "growth"]);

export function keywords(text: string): Set<string> {
  const out = new Set<string>();
  for (const w of (text || "").toLowerCase().match(/[a-z][a-z0-9-]{3,}/g) || []) {
    if (!STOP.has(w)) out.add(w);
  }
  return out;
}

export function concepts(words: Set<string>): Set<string> {
  const out = new Set<string>();
  for (const [name, vocab] of Object.entries(CONCEPTS)) {
    if (vocab.some((v) => words.has(v))) out.add(name);
  }
  return out;
}

// Ardent's own sector label leads, because it was written about this fund by
// someone deciding whether to call them. Crawled focus tags follow.
function sectorTerms(inv: Investor): string[] {
  const out: string[] = [];
  if (inv.ardent_sector) out.push(inv.ardent_sector);
  for (const f of inv.focus || []) if (f) out.push(f);
  return out;
}

export function excludedBy(m: Mandate, inv: Investor, want: Set<string>): string | null {
  if (inv.never_approach) return "graded never-approach";
  const hard = new Set(m.hard_filters || []);

  if (hard.has("geography")) {
    // Where the fund DEPLOYS, not where it keeps its office. Known failure
    // only: an empty list is a gap in our data, not a reason to hide a fund.
    const region = regionFor(m.country_code);
    const theirs = inv.invest_geographies || [];
    if (region && theirs.length && !theirs.includes(region)) {
      return `does not invest in ${region}`;
    }
  }
  if (hard.has("fund_type") && (m.fund_types || []).length) {
    // Known failure only, as everywhere else: ~200 investors carry no fund type
    // on the sheet and are kept rather than hidden by a gap in our own record.
    if (inv.fund_type && !m.fund_types.includes(inv.fund_type)) {
      return `not ${m.fund_types.map(fundTypeLabel).join("/")}`;
    }
  }
  if (hard.has("sector") && want.size) {
    const terms = sectorTerms(inv);
    if (terms.length) {
      const theirs = concepts(keywords(terms.join(" ")));
      if (![...theirs].some((c) => want.has(c))) return "sector mismatch";
    }
  }
  if (hard.has("size")) {
    const target = m.expected_ev_gbp ?? m.revenue_gbp;
    if (target && inv.cheque_min != null && inv.cheque_max != null) {
      if (target < Number(inv.cheque_min) || target > Number(inv.cheque_max)) {
        return "cheque range excludes the mandate";
      }
    }
  }
  return null;
}

export type Scored = Investor & {
  stated: number;
  revealed: number | null;
  reasons: Record<string, string>;
};

export function score(
  m: Mandate, inv: Investor, want: Set<string>, wantConcepts: Set<string>
): Scored {
  const stated: Record<string, number> = {};
  const revealed: Record<string, number> = {};
  const reasons: Record<string, string> = {};

  const terms = sectorTerms(inv);
  if (terms.length && wantConcepts.size) {
    const theirs = concepts(keywords(terms.join(" ")));
    const shared = [...wantConcepts].filter((c) => theirs.has(c));
    stated.sector_fit = shared.length / wantConcepts.size;
    if (shared.length) {
      const matched = terms.filter((t) =>
        [...concepts(keywords(t))].some((c) => shared.includes(c)));
      const src = inv.ardent_sector && matched.includes(inv.ardent_sector)
        ? "Ardent records focus on" : "states focus on";
      reasons.sector_fit = `${src} ${matched.slice(0, 3).join(", ")}`;
    }
  }

  if ((m.fund_types || []).length) {
    stated.deal_type_fit = inv.fund_type && m.fund_types.includes(inv.fund_type) ? 1 : 0;
    if (stated.deal_type_fit === 1) {
      reasons.deal_type_fit = `${fundTypeLabel(inv.fund_type)} on the Ardent list`;
    }
  }

  const target = m.expected_ev_gbp ?? m.revenue_gbp;
  if (target && (inv.cheque_min != null || inv.cheque_max != null)) {
    const lo = Number(inv.cheque_min ?? 0);
    const hi = Number(inv.cheque_max ?? lo * 10);
    const band = inv.cheque_source === "ardent_band" ? inv.check_band : null;
    if (lo <= target && target <= hi) {
      stated.size_fit = 1;
      reasons.size_fit = band
        ? `Ardent bands them at ${band}, mandate at ${(target / 1e6).toFixed(0)}m`
        : `writes ${(lo / 1e6).toFixed(0)}–${(hi / 1e6).toFixed(0)}m, mandate at ${(target / 1e6).toFixed(0)}m`;
    } else {
      // Orders of magnitude, so 2x out is not the same as 50x out.
      const edge = target < lo ? lo : hi;
      const ratio = Math.max(target, edge) / Math.max(Math.min(target, edge), 1);
      stated.size_fit = Math.max(0, 1 - (ratio - 1) / 4);
    }
  }

  const region = regionFor(m.country_code);
  const theirs = inv.invest_geographies || [];
  if (region && theirs.length) {
    stated.geography_fit = theirs.includes(region) ? 1 : 0;
    if (theirs.includes(region)) reasons.geography_fit = `invests in ${theirs.join(", ")}`;
  } else if (m.country_code && inv.country_code) {
    // No deployment regions on file — fall back to the office, and say so by
    // scoring it lower than a stated match ever gets.
    stated.geography_fit = inv.country_code === m.country_code ? 0.6 : 0.2;
    if (inv.country_code === m.country_code) {
      reasons.geography_fit = `office in ${inv.country_code} (no investment geography on file)`;
    }
  }

  if (inv.blob) {
    const theirWords = keywords(inv.blob);
    const shared = [...want].filter((w) => theirWords.has(w));
    if (want.size) {
      revealed.thesis_similarity = Math.min(1, (shared.length / want.size) * 2);
      if (shared.length) {
        reasons.thesis_similarity =
          `portfolio overlaps on ${shared.slice(0, 5).join(", ")}`;
      }
    }
  }
  if (inv.holdings > 0) {
    revealed.recent_activity = Math.min(1, (inv.recent || 0) / 5);
    if (inv.latest) {
      reasons.recent_activity =
        `${inv.recent} investments in the last 4 years, most recent ${inv.latest}`;
    }
  }

  // Divide by ALL weights: a missing signal is unknown, not neutral, so
  // knowing three things about an investor beats knowing one.
  const blend = (s: Record<string, number>, w: Record<string, number>) => {
    if (!Object.keys(s).length) return 0;
    const total = Object.values(w).reduce((a, b) => a + b, 0);
    return Object.entries(w).reduce((acc, [k, weight]) => acc + (s[k] || 0) * weight, 0) / total;
  };

  return {
    ...inv,
    stated: Number(blend(stated, weightsFor(m)).toFixed(4)),
    revealed: Object.keys(revealed).length
      ? Number(blend(revealed, REVEALED).toFixed(4))
      : null,
    reasons,
  };
}
